drivers/flash/flash_ifx_cat1.c

```diff
@@ -0,0 +1,190 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT	  infineon_cat1_flash_controller
+#define SOC_NV_FLASH_NODE DT_PARENT(DT_INST(0, fixed_partitions))
+
+#define PAGE_LEN DT_PROP(SOC_NV_FLASH_NODE, erase_block_size)
+
+#include <zephyr/kernel.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/flash.h>
+#include <zephyr/logging/log.h>
+
+#include "cyhal_flash.h"
+
+LOG_MODULE_REGISTER(flash_infineon_cat1, CONFIG_FLASH_LOG_LEVEL);
+
+/* Device config structure */
+struct ifx_cat1_flash_config {
+	uint32_t base_addr;
+	uint32_t max_addr;
+};
+
+/* Data structure */
+struct ifx_cat1_flash_data {
+	cyhal_flash_t flash_obj;
+	struct k_sem sem;
+};
+
+static struct flash_parameters ifx_cat1_flash_parameters = {
+	.write_block_size = DT_PROP(SOC_NV_FLASH_NODE, write_block_size),
+	.erase_value = 0x00,
+};
+
+static inline void flash_ifx_sem_take(const struct device *dev)
+{
+	struct ifx_cat1_flash_data *data = dev->data;
+
+	k_sem_take(&data->sem, K_FOREVER);
+}
+
+static inline void flash_ifx_sem_give(const struct device *dev)
+{
+	struct ifx_cat1_flash_data *data = dev->data;
+
+	k_sem_give(&data->sem);
+}
+
+static int ifx_cat1_flash_read(const struct device *dev, off_t address, void *buffer, size_t length)
+{
+	struct ifx_cat1_flash_data *data = dev->data;
+	const struct ifx_cat1_flash_config *config = dev->config;
+
+	flash_ifx_sem_take(dev);
+
+	cy_rslt_t rslt = cyhal_flash_read(&data->flash_obj, address + config->base_addr,
+					  (uint8_t *)buffer, length);
+	if (rslt != CY_RSLT_SUCCESS) {
+		flash_ifx_sem_give(dev);
+		return -EIO;
+	}
+	flash_ifx_sem_give(dev);
+	return 0;
+}
+
+static int ifx_cat1_flash_write(const struct device *dev, off_t offset, const void *data,
+				size_t data_len)
+{
+	struct ifx_cat1_flash_data *dev_data = dev->data;
+	const struct ifx_cat1_flash_config *dev_config = dev->config;
+	uint32_t write_offset = dev_config->base_addr + (uint32_t)offset;
+	uint8_t *ptr = (uint8_t *)data;
+
+	if (!data_len || (offset < 0) || (data_len % PAGE_LEN != 0) || (offset % PAGE_LEN != 0)) {
+		return -EINVAL;
+	}
+
+	flash_ifx_sem_take(dev);
+
+	while (data_len) {
+		cy_rslt_t rslt =
+			cyhal_flash_write(&dev_data->flash_obj, write_offset, (uint32_t *)ptr);
+		if (rslt != CY_RSLT_SUCCESS) {
+			LOG_ERR("Error in writing @ 0x%x (Err:0x%x)", write_offset, rslt);
+			return -EIO;
+		}
+
+		ptr += PAGE_LEN;
+		write_offset += PAGE_LEN;
+		data_len -= PAGE_LEN;
+	}
+
+	flash_ifx_sem_give(dev);
+
+	return 0;
+}
+
+static int ifx_cat1_flash_erase(const struct device *dev, off_t offset, size_t size)
+{
+	struct ifx_cat1_flash_data *data = dev->data;
+	const struct ifx_cat1_flash_config *config = dev->config;
+	uint32_t erase_offset = config->base_addr + (uint32_t)offset;
+	cy_rslt_t rslt;
+
+	if ((offset > 0) && ((offset % PAGE_LEN) != 0)) {
+		return -EINVAL;
+	}
+
+	if (((erase_offset + size) < config->max_addr) && ((size % PAGE_LEN) != 0)) {
+		return -EINVAL;
+	}
```

Comment:
`(erase_offset + size) < config->max_addr)` isn't this actually good?
Shouldn't `&&` be `||`?